### DISCUSSION TITLE
Expose real-time PP as Bindable

### DISCUSSION THREAD

**Discussion by nahkd123** [ppy/osu#15531] [Ideas] [answered]
Currently projects that uses ``ppy.osu.Game`` package have to either write the entire real-time PP calculator code, or use reflection to access ``PerformancePointsCounter`` inside ``SkinnableTargetContainer`` (which is not a good idea, but it works)

By exposing the real-time PP value as Bindable, other projects can access to it without hacky methods. I would suggest putting the Bindable inside ``ScoreProcessor``

I want to create PR for this, but to be honest, the codebase looks like a maze to me. So I guess I'll have to ask for this idea to be implemented.

**Comment by peppy** [comment, marked as answer]
What you are saying makes very little sense, so maybe it would be better if you start by explaining what you are attempting to do.

**Reply by nahkd123** [reply]
Basically I'm trying to make live performance points graph HUD for my replay viewer, which involves reading real-time PP value.

**Reply by peppy** [reply]
osu! already shows pp values during replay playback, can you describe further what change you're trying to make?

Please provide code if possible, it may better explain your intentions.

**Reply by peppy** [reply]
Okay so it sounds like you're making a third party app hacking pieces together to make it work.

If you are actually rendering a `Player` instance and just want to get the exact value being displayed, you can use `ChildrenOfType` to find the HUD element you need. No reflection is required.

If you're rendering gameplay yourself, use `GetTimedDifficultyAttributesAsync` which is already exposed. There's no need to use `PerformancePointsCounter` if you're reimplementing from scratch.

**Reply by nahkd123** [reply]
I'm trying to bring the entire real-time PP calculation to ``ScoreProcessor`` so that I can use it in my performance points graph HUD, something like this:

![image](https://user-images.githubusercontent.com/20136708/140854617-c08b8546-8876-404d-8393-6132d464c70e.png)
(yes it's upside down)

Currently I have to use huge chunk of code and reflections to access the real-time PP value, something like this:
```cs
            PerformanceGraph performanceGraph;
            AddInternal(performanceGraph = new()
            {
                Anchor = Anchor.TopLeft,
                Origin = Anchor.TopLeft,
                Width = 300,
                Margin = new MarginPadding { Left = 10f, Top = 50f }
            });

            var skinnableTarget = HUDOverlay.GetRecursive(v => v is SkinnableTargetContainer) as SkinnableTargetContainer;

            // Hacky way to get live PP
            DrawableRuleset.Playfield.NewResult += (dho, judgement) =>
            {
                if (ppCounter == null)
                {
                    var internalChild = DrawablesUtils.GetInternalChildren(skinnableTarget);
                    var containerList = internalChild.OfType<SkinnableTargetComponentsContainer>();
                    if (containerList.Any()) foreach (var a in containerList.First().Children) if (a is PerformancePointsCounter ppCounter)
                    {
                        performanceGraph.Counter = ppCounter;
                    }
                }
            };
```

**Reply by peppy** [reply]
So if you want to improve your code you can just do

`HUDOverlay.ChildrenOfType<PerformancePointsCounter>.First().Current`

This is your bindable. It can replace 100% of the code you have, including the "new result" binding.

If you want to be independent of the counter, you can use `GetTimedDifficultyAttributesAsync` directly.

I don't believe we want score processor to calculate PP as in your proposal. It would be a new processor if anything. Score and PP are different tasks to us.

**Reply by smoogipoo** [reply]
We'll need some temporary cache for that `GetTimedDifficultyAttributes` method at some point, because currently it does a full calc every time. Maybe something that lasts only as long as the beatmap is loaded/in the working beatmap cache.
